contrib: verify torrents with verify-binary.py

willcl-ark opened the pull request:
Closes #27702

Adds functionality to the verify-binary.py script to verify the signatures in a torrent download.
Users must download the torrent themselves manually and point the script to the directory using the `torrent` subcommand.

The syntax chosen was `./contrib/verify-binaries/verify.py torrent <path to SHA256SUMS file>`

willcl-ark committed:
contrib: refactor verify-binary print output

Sjors commented on contrib/verify-binaries/verify.py:
```diff
@@ -459,14 +460,17 @@ def verify_binary_hashes(hashes_to_verify: t.List[t.List[str]]) -> t.Tuple[Retur
-    bad: t.List[SigData],
-    files_to_hashes: t.Dict[str, str],
-    missing_files=None,
-    ):
```
a6dcf2cb0817143beacf0b9004191a0dd6d2e43c Did you mean to change the indentation here? Maybe do so in the first refactor commit?

willcl-ark replied:
Thanks fixed in next push

Sjors commented on contrib/verify-binaries/verify.py:
```diff
@@ -481,9 +485,8 @@ def print_output(
-        if missing_files:
-            for filename in missing_files:
-                print(f"MISSING: {filename}")
+        for filename in missing_files:
```
a6dcf2cb0817143beacf0b9004191a0dd6d2e43c: did you mean to do this in the refactor commit?

willcl-ark replied:
Thanks fixed in next push

Sjors reviewed (commented):
Successfully tested on macOS 13.4 against v25.0.

Lightly code reviewed a6dcf2cb0817143beacf0b9004191a0dd6d2e43c.

Do we want to use the .asc file from the torrent, or should we just get it from the Guix signature repo? The latter may have more signatures, as they keep coming in after release.

willcl-ark committed:
contrib: verify binary from torrent download

Requires torrent to be downloaded externally to this script.

kristapsk commented:
> Do we want to use the .asc file from the torrent, or should we just get it from the Guix signature repo? The latter may have more signatures, as they keep coming in after release.

It's pros and cons. Repo will have more signatures, but any centralized server is single point of failure, torrents are better in this regard.

willcl-ark commented:
Sure there are tradeoffs.

It also seems problematic in that fetching them in any sane way would  require `git` to be installed on the host machine (so we could e.g. do a sparse checkout of the release sub-directory). Alternatively just download the entire repo as .zip and go from there? Both seem undesirable/"messy" to me personally. 🤷🏼‍♂️

Sjors commented:
In the case of Github, I believe you can get individual raw files too.

But there's something to be said for not being dependent on having Git installed, and also have the option to not download if you used a torrent before. So maybe we can have a `--signatures-from-repo` like option in that case? For non-torrent imo this should be `true` by default if git is installed. For torrents the option could be suggested.

willcl-ark committed:
contrib: verify-binaries.py fetch additional sigs

from guix.sigs repo

Enabled automatically for `pub` and `bin`, but disabled for `torrent` to
for maximum privacy.

Can be toggled using the --signatures-from-repo option.

willcl-ark commented:
OK I implemented as you suggested @Sjors. It does seem like a win to get more signatures on balance.

If people feel the approach is OK, and that it belongs in this PR (and not it's own), I can probably tidy it up further as theres still some duplicated code which can probably be factored out.

willcl-ark commented on contrib/verify-binaries/verify.py:
```diff
@@ -692,16 +866,23 @@ def main():
         help="Path to a binary distribution file to verify. Can be specified multiple times for multiple files to verify."
     )
 
+    torrent_parser = subparsers.add_parser("torrent", help="Verify local torrent dir (requires manual torrent download).")
```
Right, without performing the actual torrent download it ends up being the same!

I'm going to close this for now as IMO it doesn't make sense to pull in a large python torrent lib just to perform the download in this verification script. Perhaps we close #27702 as well, unless we really want this.

If there was appetite to add a python-based torrent download, I'd be happy to work on it...
